[CT9ARyan] iP

CT9ARyan opened the pull request.

CT9ARyan committed:
Add Level-0 increment: Rename, Greet, Exit

CT9ARyan committed:
Add formatting method to print divider line

CT9ARyan committed:
Add Level-1 increment: Echo

CT9ARyan committed:
Add Level-2 increment: Add, List

CT9ARyan committed:
Add new Task class

CT9ARyan committed:
Add getter and setter for task name in Task class

CT9ARyan committed:
Add Level-3 increment: Mark as Done

CT9ARyan committed:
Refine if-else block for "mark" and "unmark" commands

CT9ARyan committed:
Add exception handling for invalid inputs for "mark" and "unmark" commands and fix some minor bugs

Change the to-be-compared string used from "keyword" to "line" for "bye" and "list" commands so that they will perform their corresponding actions only when the user enters these exact commands. This improves flexibility and allows the user to reuse these keywords in their other tasks/items for the list.

CT9ARyan committed:
Edit code to conform to java coding standards.

CT9ARyan committed:
Add Todo class

CT9ARyan committed:
Add Deadline class

CT9ARyan committed:
Add Event class

CT9ARyan committed:
Modify Task class

Change access modifiers for attributes from private to protected so that subclasses inheriting from Task class can access these attributes. Add extra methods defined in the subclasses so that the subclasses can override them. This is useful when the subclass objects are being referenced/type-casted as a superclass object. By adding these extra methods in the superclass and overridding them in the subclass, it ensures that the overridden methods from the subclass will be invoked instead of the one in the superclass, otherwise the methods found only in the subclass can no longer be invoked, leading to an error.

CT9ARyan committed:
Add Level-4 increment: Todo, Event, Deadline

Modify Atom.java to allow user to specify, store and track 3 types of tasks, namely todo, deadline and event. By specifying the type of task before adding it to the list, the list now displays more important info, like the type of task, the deadline of the task, the timeline of the event, etc. depending on the type of task. In order to achieve this, the contents of the printList() method are also changed. Constants are declared at the top and used whenever applicable so that they make sense, resulting in improved code readability, but most importantly it gets rid of any magic numbers.

kq2003 commented on src/main/java/Atom.java:
```diff
@@ -0,0 +1,167 @@
+        String[] words = line.split(" ");
+        String keyword = words[0].toLowerCase();
+
+        while (!line.equalsIgnoreCase("bye")) {
```
You could've added annotations for different chunks of codes.

kq2003 commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,19 @@
+    }
+
+    @Override
+    public String getBy() {
```
Your naming is exquisite!

kq2003 commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,19 @@
+        return by;
+    }
+
+    @Override
```
This is good coding standard!

kq2003 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,48 @@
+    public Task(String item) {
+        this.item = item;
+        this.isDone = false;
+        taskCount++;
```
++ is good!

kq2003 replied:
I like your use of ++!

kq2003 reviewed (commented):
I really like how you followed the coding standards!

kennethszj commented on src/main/java/Atom.java:
```diff
@@ -0,0 +1,167 @@
+                if (Task.getTaskCount() != 0) {
+                    printList(Arrays.copyOf(tasksList, Task.getTaskCount()));
+                } else {
+                    System.out.println("Oh oh! List is empty.");
```
try to avoid deep nesting

kennethszj commented on src/main/java/Atom.java:
```diff
@@ -0,0 +1,167 @@
+        Task[] tasksList = new Task[100];
+
+        System.out.print("Enter command: ");
+
```
good job as Array specifiers must be attached to the type not the variable.

kennethszj commented on src/main/java/Atom.java:
```diff
@@ -0,0 +1,167 @@
+import java.util.Scanner;
+import java.util.Arrays;
+
+public class Atom {
```
good effort in avoiding magic numbers

kennethszj commented on src/main/java/Atom.java:
```diff
@@ -0,0 +1,167 @@
+
+        String line;
+        Scanner scanner = new Scanner(System.in);
+        Task[] tasksList = new Task[100];
```
try not to use magic numbers

kennethszj reviewed (commented):
overall is readable with minor violations in code quality

CT9ARyan committed:
Update and refactor code to improve code quality

Firstly, by adopting the rule of "make the happy path prominent", I managed to simplify some of the if blocks in the code by switching the order of the logic and adding guard clauses. This helped to remove the redundant else block, making the if block a lot neater and easier to read. Secondly, my initial iteration of the code encountered an issue of deep nesting, aka the arrowhead style code, and this was the biggest (visual) flaw in my code. By refactoring my code, I managed to extract the contents in the large if block into 2 separate methods. This resulted in a short, concise and neat code which is still easy to understand and read. Lastly, I also refactored the "magic numbers" as constants, declared at the top of the code. This gives meaning to these arbituary values that I used and helps other readers make sense of them when they read through my code. Next, I also added exception handling functionalities for both the "deadline" and "event" commands in the event that the user enters an incorrect command format. Lastly, I created a method to print out the default exception message.

CT9ARyan committed:
Use TextUI testing to semi automate the testing process and check for potential regressions

Using the input/output redirection technique, the input.txt stores all the input that are to be fed to the program/source code, and the output will be stored in another txt file called "ACTUAL.txt". The latter is then compared with EXPECTED.txt - a file containing the expected output of the code, given the inputs defined in the input.txt file, and any difference in the files will be displayed in the terminal.

Initially, even after following all the steps correctly and ensuring that the contents in both EXPECTED.txt and ACTUAL.txt LOOK exactly the same, the file compare (FC) command still showed a difference in the 2 files. To resolve this, I had to open these 2 files in my Notepad++ app and highlight all the contents of the file. Narrowing down my search to the lines where the errors were reported, this was when i realised that the difference was actually empty space characters, which cannot be seen by the naked eye. Following that, I went back to my code to make the necessary changes, and now the files are identical!

In summary, this exercise taught me how to semi automate the testing of textUIs using the I/O redirection technique. This tool/skill proved to be effective as it helps to check for potential bugs/regressions whenever we made changes to the code, and this makes the whole testing process easier, instead of having to manually write the test cases.

CT9ARyan committed:
Minor bug fix in if else block for "list" command

The use of the return keyword in the if block exits the whole outer while loop, leading to the termination of the program when the list is empty (i.e when Task.getTaskCount() == 0). Therefore, to rectify this, the else keyword is used instead when the list is empty.

CT9ARyan committed:
Update .gitignore

CT9ARyan committed:
Add custom exception class to handle empty task id for "mark" and "unmark" commands

CT9ARyan committed:
Add custom exception class to handle invalid (out-of-bounds) indexing of the task list during "mark" or "unmark" operations

CT9ARyan committed:
Add custom exception class to handle empty TODO task description

CT9ARyan committed:
Add custom exception class to handle empty DEADLINE task description

CT9ARyan committed:
Add custom exception class to handle invalid formatting of the DEADLINE command

Some examples of invalid formatting include, using "by" instead of "/by" or not using the keyword "by".

CT9ARyan committed:
Add custom exception class to handle empty EVENT task description

CT9ARyan committed:
Add custom exception class to handle invalid formatting of the EVENT command

The implementation of this is quite similar to that of DEADLINE. However, for the EVENT command, there are now 2 keywords to check for now (i.e "/from" and '/to") instead of just 1. Therefore, this exception class handles incorrect inputs such as "from" or "to" without the "/", and empty FROM or TO date/time inputs.

CT9ARyan committed:
Update Atom to handle errors like incorrect user inputs

Using all the custom exception classes defined earlier, Atom is now able to handle errors and exceptions like incorrect user inputs. For the exception handling, a "try-catch" block, together with the "throw" keyword is implemented for every command block in Atom, so that every command block can throw and catch the corresponding exception depending on the user input. Multiple catch blocks are used if the exception handler in the command block needs to catch and handle different types of exception.

Furthermore, to improve on code readability, I refactored the code by extracting 3 different methods from the if-else command blocks in the main method, namely "addTodoTask", "addDeadlineTask", and "addEventTask". This makes the main part of the code easier to read, understand, manage and it also looks more visually pleasing. I also refactored the size of the task list from "100" to "MAX_NUMBER_OF_TASKS" so as to avoid using magic numbers.

CT9ARyan committed:
Merge branch 'branch-Level-5'

CT9ARyan committed:
Organise classes into packages

Created 3 new packages.

1. "atom" --> main package/folder storing all the classes and packages related to Atom
2. "exception" --> stores all of Atom's custom exception classes
3. "task" --> stores all the task related classes (i.e Task, Todo, Deadline, Event)

Following that, the package statment is included at the top of every source file. This was achieved easily thanks to the refactoring feature in the IDE. In addition, in the main Atom source code, all the classes residing in the other packages need to be imported so that they can be used in the main Atom source code. I also ensured that all these new additions conformed to the coding standards, like ensuring all the packages' name are in all lower case, and when importing all the classes used in a package, they have to be stated explicitly instead of using the * operator.

CT9ARyan committed:
Merge branch 'branch-A-Packages'

CT9ARyan committed:
Update code to use ArrayList instead of Arrays.

CT9ARyan committed:
Remove static variable "taskCount" and its methods in Task class

Since Atom is now using an ArrayList instead of an Array to store the Tasks, the built-in method .size() of the ArrayList class can be used to get the size/number of tasks in the list instead. This removes the need for an extra static variable and getter method to read and query the number of current tasks in the list. Doing so makes the code a lot neater, easier and more straightforward too.

CT9ARyan committed:
Update printList() method

The if condition that checks whether the list is empty was initially written in the main() method. This if block was moved to the printList() method so that the code in the main() looks more concise. Furthermore, it makes more sense to put it in the printList() method as it should first perform a check to ensure that the list is not empty before actually printing the contents of the list.

CT9ARyan committed:
Update display format for DEADLINE and EVENT tasks after "mark" and "unmark" commands

Initially, after marking/unmarking a DEADLINE or EVENT task, the output will show the type of task and the name of the task, without any additional information. After the change, the output now not shows the type of task and the name of the task, it also shows the date and time the task has to be completed by for a DEADLINE task, or the timeline or duration of the event for an EVENT task.

CT9ARyan committed:
Add Level-6 increment: Delete

Atom is now able to delete tasks from the list that are no longer needed. A method deleteTask() is created to delete the specified task from the list and print a message to the user, showing the user the deleted task and the number of tasks remaining in the list. In the main() method, similar to the "mark" and "unmark" commands, all the exceptions like incorrect format and missing task id are handled properly and gracefully, with the appropriate error messages. In addition, as mentioned before, all instances of the static variable "taskCount" are now replaced with tasksList.size().

CT9ARyan committed:
Import file and file exception classes

CT9ARyan committed:
Create new folder "data" and txt file "AtomList" in the "data" folder

CT9ARyan committed:
Save list data by writing to AtomList.txt file.

CT9ARyan committed:
Load data from AtomList.txt file to the list when the program starts

CT9ARyan committed:
Merge branch 'branch-Level-6'

CT9ARyan committed:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/atom/Atom.java

CT9ARyan committed:
Edit methods for adding a TODO, DEADLINE and EVENT tasks

Initially, the adding of a TODO, DEADLINE and EVENT task creates and adds the specific task object, based on its type to the list, and prints out a "task added" message all in the same method. With the implementation of the Level-7 increment: Save, these methods need to be reused in order to add the task's data stored in the file to the list. However, as the tasks are added from the txt file to the list at the start of the program, all the corresponding task messages will also be printed at the start, flooding the console terminal with loads of "task added" messages, which are not required to be shown to the user. Therefore, I have further refactored the method, essentially separating it into 2 different parts. (i.e one with message and one without)

Currently, the addTask() method solely creates and adds the task object to the list, while the addTaskWithMessage() method prints out the "task added" message, on top of adding the task using the addTask() method.

tzerbin commented on src/main/java/atom/Atom.java:
```diff
@@ -0,0 +1,441 @@
+    public static final int TO_START_INDEX_OFFSET = 4;
+
+    public static void printDivider() {
+        System.out.println("__________________________________________________");
```
Please refrain from using magic strings and numbers in the entire codebase.

tzerbin commented on src/main/java/atom/Atom.java:
```diff
@@ -0,0 +1,441 @@
+
+public class Atom {
+
+    public static final int TODO_START_INDEX = 5;
```
Good usage of constants!

tzerbin commented on src/main/java/atom/Atom.java:
```diff
@@ -0,0 +1,441 @@
+        System.out.println("You now have " + tasksList.size() + " tasks in your list.");
+    }
+
+    public static void main(String[] args) {
```
The main method is long, perhaps abstract out code into smaller methods?

tzerbin commented on src/main/java/atom/task/Task.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    public String setTaskType() {
+        return "-";
```
Perhaps use enums instead of a string to represent the type of tasks?

tzerbin reviewed (commented):
Good attempt on following guidelines!

The code base is very long in main method, consider abstracting chunks of code into methods.

CT9ARyan committed:
Update .gitignore

CT9ARyan committed:
Add new class "TaskList" to handle all task related operations

This class contatins:
- task list
- method to get the task list
- method to get the task list size
- methods to mark or unmark a certain task
- methods to add different types of task to the list
- method to delete a task from the list

CT9ARyan committed:
Add new class "Ui" to handle all interactions with the user

This class handles the display format of the application, reads the inputs from the user, and prints the relevant messages/output for the user depending on the commands given.

CT9ARyan committed:
Add new class "Storage" to handle transfer of saved data between AtomList.txt and the ArrayList in Atom

This class handles:
- loading of saved data (if any) from the AtomList.txt file to the ArrayList in Atom
- storing of data from the ArrayList in Atom to the AtomList.txt file upon exiting the program.

CT9ARyan committed:
Add new class "Parser" to parse and make sense of the user input

CT9ARyan committed:
Add new abstract class "Command"

This class, which will be the superclass, defines the methods that need to be implemented by its subclasses (i.e the specific command classes)

CT9ARyan committed:
Add new class "ExitCommand" which inherits from Command class

This command is used to:
- exit the program
- store the saved data from the ArrayList to the AtomList.txt file

CT9ARyan committed:
Add new class "ListCommand" which inherits from Command class

This command prints the contents of the ArrayList to the console terminal.

CT9ARyan committed:
Add new class "InvalidCommand" which inherits from Command class

CT9ARyan committed:
Add new class "MarkCommand" which inherits from Command class

This command is used to:
- mark a task as done
- mark a task as undone

CT9ARyan committed:
Add new class "AddCommand" which inherits from Command class

This command can:
- add a Todo task
- add a Deadline task
- add a Event task

CT9ARyan committed:
Add new class "DeleteCommand" which inherits from Command class

This command is used to:
- delete a task from the list, given the task id

CT9ARyan committed:
Add new exception class specific to Atom "AtomException"

CT9ARyan committed:
Refactor variable names to improve code quality

For deadline related tasks:
String by -> String dueDate

For event related tasks:
String from -> String startDate
String to -> String endDate

This simple yet effective change makes the code easier to understand and read. Lastly, the "Task class" is updated to be an abstract class.

CT9ARyan committed:
Update all exception classes to store the corresponding error messages

These error messages are stored in an overridden method called "getMessage()". This makes the retrieval of the error messages easier (using e.getMessage() )

CT9ARyan committed:
Refactor code to make it more OOP

CT9ARyan committed:
Organise related classes in their respective packages and update all import statements

CT9ARyan committed:
Add new class "DateTimeParser" to parse the date and time entered by user for DEADLINE and EVENT tasks

CT9ARyan committed:
Add method in "DateTimeParser" class to validate date and time entered by user

This method first checks whether the date time format in the user command (DEADLINE and EVENT) is valid and throws the corresponding exception if it's not. Next, it checks for the 5 different parameters in the date time input. They are "day", "month" and "year" for DATE and "hour" and "min" for TIME. If any of the parameters are invalid (non-integers or contain extra non-integer characters), the parse method will throw an exception and hence the input is deemed invalid.

CT9ARyan committed:
Add methods in "DateTimeParser" class to check rationality of each parameter's value

CT9ARyan committed:
Add method in "DateTimeParser" class to parse a valid date time input entered by user and display it in a more readable format in the list

CT9ARyan committed:
Add methods to print error messages regarding invalid date time inputs

CT9ARyan committed:
Fix a minor bug and update methods name

Minor bug:
Initially, in the "DeleteCommand" class, after deleting the task from the list, there was an IndexOutOfBoundsException as the "showDeleteTaskMessage" method declared in the "Ui" class needed to access the index of the deleted item, to print out the message. Therefore, to resolve this issue, the message is printed first before deleting the task from the list. As such, the "tasks.deleteTask(taskId)" line is shifted down. To ensure correctness, since the message has to be printed first before the deletion of the task from the list, the size of the Arraylist is pre-decremented by 1.

CT9ARyan committed:
Import the "DateTimeParser" class to the "AddCommand" class and update code logic for DEADLINE and EVENT commands

CT9ARyan committed:
Add new class "FindCommand" which inherits from Command class

CT9ARyan committed:
Update "Parser" class to handle the "FindCommand" case

CT9ARyan committed:
Add new method in "TaskList" class to find matching tasks containing the keyword in the list

CT9ARyan committed:
Refactor "printList" method and add new method to print list of matching tasks found

The "printList" method was refactored by extracting part of it to become another method. Initially, the "printList" method contains the code for printing out messages to the terminal and the for loop that iterates through the list and prints out all the tasks in the list. After refactoring, the latter part was extracted to another method, named "printTasksInList". This not only makes the "printList" method simpler and easier to read, but also improves reusability. In fact, the newly added method, which prints out another list containing the matching tasks found based on the "find " command, uses the newly extracted method.

CT9ARyan committed:
Update methods in "FindCommand" class

CT9ARyan committed:
Add new custom exception class to handle missing keyword for "find" command

CT9ARyan committed:
Change access modifiers for constants from PUBLIC to PRIVATE

CT9ARyan committed:
Merge pull request #1 from CT9ARyan/branch-Level-9

Level-9 Increment: Find

CT9ARyan committed:
Merge branch 'master' into branch-Level-8

CT9ARyan committed:
Add new class "FilterCommand" which inherits from Command class

CT9ARyan committed:
Update "Parser" class to handle the "filter" command case

CT9ARyan committed:
Add new method in "TaskList" to filter tasks by date specified by user and add them to new list

CT9ARyan committed:
Add new method to check validity of date entered by user

CT9ARyan committed:
Add new method to parse valid date into a more readable format

CT9ARyan committed:
Add new method in "Ui" class to print list of tasks filtered by the specified date

CT9ARyan committed:
Add more methods to print error messages regarding invalid dates or date format

CT9ARyan committed:
Add new custom exception class to handle missing date in "filter" command

CT9ARyan committed:
Update methods and logic in "FilterCommand" class

CT9ARyan committed:
Merge pull request #2 from CT9ARyan/branch-Level-8

Level-8 increment: Dates and Times

CT9ARyan committed:
Add JavaDoc comments for all classes in "command" package

CT9ARyan committed:
Add JavaDoc comments for all classes in "exception" package

CT9ARyan committed:
Add JavaDoc comments for all classes in "parser" package

CT9ARyan committed:
Add JavaDoc comments for "Storage" class

CT9ARyan committed:
Add JavaDoc comments for all classes in "task" package

CT9ARyan committed:
Add JavaDoc comments for "TaskList" class

CT9ARyan committed:
Add JavaDoc comments to "Ui" class

CT9ARyan committed:
Merge pull request #3 from CT9ARyan/branch-A-JavaDoc

A-JavaDoc increment: JavaDoc

CT9ARyan committed:
Add missing JavaDoc comment for "parse" method in "Parser" class

CT9ARyan committed:
Fix minor bug in "Parser" class

Initially, the ExitCommand, "bye" and ListCommand, "list" both depend on the keyword (first word/string of the user command) of the user input command. As a result, even if the user enters an errorneous input command like "bye     1" or "list     hello world", the corresponding command actions will be still taken. This is not the expected behaviour of ATOM. As such, the code logic has been changed such that only the ExitCommand and the ListCommand will depend on the full user command. The other commands will still depend on the keyword of the command.

CT9ARyan committed:
Update .gitignore

CT9ARyan committed:
Update chatbot name and add product screenshot and intro in README.md

CT9ARyan committed:
Add a list of contents for user guide in README.md

CT9ARyan committed:
Add descriptions for todo, deadline and event commands for user guide in README.md

CT9ARyan committed:
Add descriptions for list command in user guide in README.md

CT9ARyan committed:
Add descriptions for mark and unmark commands in user guide in README.md

CT9ARyan committed:
Add descriptions for delete command in user guide in README.md

CT9ARyan committed:
Add descriptions for find command in user guide in README.md

CT9ARyan committed:
Add descriptions for filter command in user guide in README.md

CT9ARyan committed:
Add descriptions for bye command in user guide in README.md

CT9ARyan committed:
Add descriptions regarding saving the task list in user guide in README.md

CT9ARyan committed:
Add a command summary table for easy reference of important commands in user guide in README.md

CT9ARyan committed:
Add files via upload

CT9ARyan committed:
Delete AtomScreenshot.png

CT9ARyan committed:
Add files via upload

CT9ARyan committed:
Update image file path in user guide in README.md

CT9ARyan committed:
Update README.md

CT9ARyan committed:
Update README.md

CT9ARyan committed:
Update README.md

CT9ARyan committed:
Update README.md

CT9ARyan committed:
Update README.md

CT9ARyan committed:
Fix bugs in README.md

1. Change image file path so that the image can be displayed on GIthub pages.

2. Add missing section link (anchor) for "Features".

CT9ARyan committed:
Merge branch 'master' of https://github.com/CT9ARyan/ip
